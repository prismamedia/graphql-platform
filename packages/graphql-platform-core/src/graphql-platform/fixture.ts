import { Entries, fromEntries } from '@prismamedia/graphql-platform-utils';
import { DepGraph } from 'dependency-graph';
import { OperationDefinitionNode, print } from 'graphql';
import { Logger } from 'winston';
import { GraphQLPlatform, GraphQLRequest } from '../graphql-platform';
import { CreateOneDataInputValue } from './operation';
import { AnyResource, Relation, SerializedFieldValue } from './resource';
import { TypeKind, WhereUniqueInputValue } from './type';

export class FixtureGraph extends DepGraph<Fixture> {}

export type FixtureFieldData = SerializedFieldValue;

export type FixtureRelationData = null | Fixture['name'];

export type FixtureComponentData = FixtureFieldData | FixtureRelationData;

export type FixtureData = {
  [componentName: string]: FixtureComponentData;
};

export type FixtureDataMap = {
  [fixtureName: string]: FixtureData;
};

export class Fixture {
  readonly logger?: Logger;
  protected id?: WhereUniqueInputValue;

  public constructor(
    readonly name: string,
    readonly resource: AnyResource,
    readonly data: FixtureData,
    protected graph: FixtureGraph,
    protected gp: GraphQLPlatform,
  ) {
    this.logger = gp.getLogger();
  }

  public toString(): string {
    return `${this.resource}.${this.name}`;
  }

  public getRelatedFixture(relation: Relation): Fixture | null {
    this.resource.getRelationMap().assert(relation);

    const relationValue = this.data[relation.name];

    if (relationValue == null) {
      if (relation.isRequired()) {
        throw new Error(`In the fixture "${this.name}", the relation "${relation}" can't be null.`);
      }
    } else if (!(typeof relationValue === 'string' && this.graph.hasNode(relationValue))) {
      throw new Error(
        `In the fixture "${this.name}", the relation "${relation}" has to reference an "${relation.getTo()}" fixture.`,
      );
    }

    return relationValue ? this.graph.getNodeData(relationValue) : null;
  }

  public getCreateMutationAST(): OperationDefinitionNode {
    const operationDefinitionNode: OperationDefinitionNode = {
      kind: 'OperationDefinition',
      operation: 'mutation',
      variableDefinitions: [
        {
          kind: 'VariableDefinition',
          variable: {
            kind: 'Variable',
            name: {
              kind: 'Name',
              value: 'data',
            },
          },
          type: {
            kind: 'NonNullType',
            type: {
              kind: 'NamedType',
              name: {
                kind: 'Name',
                value: this.resource.getMutation('CreateOne').getDataType().name,
              },
            },
          },
        },
      ],
      selectionSet: {
        kind: 'SelectionSet',
        selections: [
          {
            kind: 'Field',
            name: {
              kind: 'Name',
              value: this.resource.getMutation('CreateOne').name,
            },
            alias: {
              kind: 'Name',
              value: 'id',
            },
            arguments: [
              {
                kind: 'Argument',
                name: {
                  kind: 'Name',
                  value: 'data',
                },
                value: {
                  kind: 'Variable',
                  name: {
                    kind: 'Name',
                    value: 'data',
                  },
                },
              },
            ],
            selectionSet: this.resource
              .getFirstPublicUnique()
              .getSelectionNode(TypeKind.Output)
              .toAST(),
          },
        ],
      },
    };

    return operationDefinitionNode;
  }

  public getCreateMutationSource(): string {
    return print(this.getCreateMutationAST());
  }

  public getCreateMutationVariables(): GraphQLRequest['variableValues'] {
    return {
      data: fromEntries([
        // Fields
        ...[...this.resource.getFieldSet()].map(field =>
          !field.isFullyManaged() ? [field.name, this.data[field.name] as FixtureFieldData] : undefined,
        ),
        // Relations
        ...[...this.resource.getRelationSet()].map(relation => {
          if (!relation.isFullyManaged()) {
            const relatedFixture = this.getRelatedFixture(relation);
            if (relatedFixture) {
              return [relation.name, { connect: relatedFixture.assertId() }];
            }
          }
        }),
      ] as Entries<CreateOneDataInputValue>),
    };
  }

  public getCreateMutation(): GraphQLRequest {
    return {
      source: this.getCreateMutationSource(),
      variableValues: this.getCreateMutationVariables(),
    };
  }

  public async load(contextValue?: GraphQLRequest['contextValue']): Promise<void> {
    try {
      const { id } = await this.gp.execute<{ id: WhereUniqueInputValue }>({
        ...this.getCreateMutation(),
        contextValue,
      });

      this.id = this.resource.getInputType('WhereUnique').assert(id);
    } catch (error) {
      this.logger && this.logger.error(error);

      throw error;
    }
  }

  public assertId(): WhereUniqueInputValue {
    if (!this.id) {
      throw new Error(`The fixture "${this.name}" is not loaded yet.`);
    }

    return this.id;
  }
}
