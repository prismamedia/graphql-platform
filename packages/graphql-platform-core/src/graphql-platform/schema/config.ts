import { GraphQLOperationType } from '@prismamedia/graphql-platform-utils';
import { GraphQLFieldConfigMap, GraphQLObjectType, GraphQLSchemaConfig as BaseGraphQLSchemaConfig } from 'graphql';
import { OperationType } from '../operation';
import { ResourceMap } from '../resource/map';

export interface GraphQLSchemaConfigConfig
  extends Omit<BaseGraphQLSchemaConfig, 'mutation' | 'query' | 'subscription'> {
  resourceMap: ResourceMap;
  queryMap: GraphQLFieldConfigMap<any, any>;
  mutationMap: GraphQLFieldConfigMap<any, any>;
}

export class GraphQLSchemaConfig implements BaseGraphQLSchemaConfig {
  public constructor(readonly config: GraphQLSchemaConfigConfig) {}

  public getOperationTypeFieldMap<TType extends OperationType>(type: TType): GraphQLFieldConfigMap<any, any> {
    const fields: GraphQLFieldConfigMap<any, any> = {};

    for (const resource of this.config.resourceMap.values()) {
      for (const operation of resource.getOperations(type)) {
        if (operation.isPublic()) {
          fields[operation.name] = operation.getGraphQLFieldConfig();
        }
      }
    }

    return fields;
  }

  public get mutation() {
    return new GraphQLObjectType({
      name: 'Mutation',
      fields: { ...this.getOperationTypeFieldMap(GraphQLOperationType.Mutation), ...this.config.mutationMap },
    });
  }

  public get query() {
    return new GraphQLObjectType({
      name: 'Query',
      fields: { ...this.getOperationTypeFieldMap(GraphQLOperationType.Query), ...this.config.queryMap },
    });
  }

  public get subscription() {
    return undefined;
  }
}
