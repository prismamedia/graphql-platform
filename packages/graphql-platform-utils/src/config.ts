export * from './config/name.js';
export * from './config/optional-deprecation.js';
export * from './config/optional-description.js';
export * from './config/optional-flag.js';
