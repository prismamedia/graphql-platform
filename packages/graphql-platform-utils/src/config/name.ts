import * as graphql from 'graphql';
import { UnexpectedValueError } from '../error.js';
import type { Path } from '../path.js';

/**
 * A string compliant with the GraphQL specification
 *
 * @see https://spec.graphql.org/draft/#sec-Names
 */
export type Name = string;

/**
 * @see https://spec.graphql.org/draft/#sec-Names
 */
export function assertName(
  maybeName: unknown,
  path: Path,
): asserts maybeName is Name {
  if (typeof maybeName !== 'string' || !maybeName) {
    throw new UnexpectedValueError(`a non-empty string`, maybeName, {
      path,
    });
  }

  try {
    graphql.assertName(maybeName);
  } catch (error) {
    throw new UnexpectedValueError(
      'to be valid against the GraphQL "Names" specification (@see: https://spec.graphql.org/draft/#sec-Names)',
      maybeName,
      { cause: error, path },
    );
  }
}

/**
 * @see https://spec.graphql.org/draft/#sec-Names
 */
export function ensureName(maybeName: unknown, path: Path): Name {
  assertName(maybeName, path);

  return maybeName;
}
