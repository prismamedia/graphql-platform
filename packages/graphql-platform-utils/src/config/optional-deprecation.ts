import { UnexpectedValueError } from '../error.js';
import { isNil, Nillable } from '../nil.js';
import type { Path } from '../path.js';

export type OptionalDeprecation = Nillable<boolean | string>;

export function assertOptionalDeprecation(
  maybeOptionalDeprecation: unknown,
  path: Path,
): asserts maybeOptionalDeprecation is OptionalDeprecation {
  if (
    !isNil(maybeOptionalDeprecation) &&
    typeof maybeOptionalDeprecation !== 'boolean' &&
    typeof maybeOptionalDeprecation !== 'string'
  ) {
    throw new UnexpectedValueError(
      `a boolean or a string`,
      maybeOptionalDeprecation,
      { path },
    );
  }
}

export function ensureOptionalDeprecation(
  maybeOptionalDeprecation: unknown,
  path: Path,
): OptionalDeprecation {
  assertOptionalDeprecation(maybeOptionalDeprecation, path);

  return maybeOptionalDeprecation;
}

export function getOptionalDeprecation(
  maybeOptionalDeprecation: OptionalDeprecation,
  defaultValue: string,
  path: Path,
): string | undefined {
  const optionalDeprecation = ensureOptionalDeprecation(
    maybeOptionalDeprecation,
    path,
  );

  return optionalDeprecation === true
    ? defaultValue
    : optionalDeprecation || undefined;
}
