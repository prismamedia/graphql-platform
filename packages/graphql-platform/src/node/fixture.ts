import * as utils from '@prismamedia/graphql-platform-utils';
import { Memoize } from '@prismamedia/ts-memoize';
import type { ConnectorInterface } from '../connector-interface.js';
import type { Node, NodeValue } from '../node.js';
import type { Seeding } from '../seeding.js';
import type { UniqueConstraintValue } from './definition.js';
import { EdgeCreationInputAction } from './type/input/creation/field.js';

export type NodeFixtureReference = string;

export type NodeFixtureData = utils.PlainObject;

export class NodeFixture<
  TRequestContext extends object,
  TConnector extends ConnectorInterface,
> {
  public readonly dependencies: ReadonlySet<NodeFixtureReference>;

  public constructor(
    public readonly seeding: Seeding<TRequestContext, TConnector>,
    public readonly node: Node<TRequestContext, TConnector>,
    public readonly reference: NodeFixtureReference,
    public readonly data: NodeFixtureData,
  ) {
    if (typeof reference !== 'string') {
      throw new utils.UnexpectedValueError('a string', reference);
    }

    if (!utils.isPlainObject(data)) {
      throw new utils.UnexpectedValueError('a plain-object', data);
    }

    this.dependencies = new Set(
      Array.from(this.node.edgesByName.values()).reduce<NodeFixtureReference[]>(
        (dependencies, edge) => {
          const maybeEdgeReference = data[edge.name];
          if (maybeEdgeReference != null) {
            dependencies.push(maybeEdgeReference);
          }

          return dependencies;
        },
        [],
      ),
    );
  }

  @Memoize()
  public get dependents(): ReadonlySet<NodeFixtureReference> {
    return new Set(
      Array.from(this.seeding.fixturesByReference.values()).reduce<
        NodeFixtureReference[]
      >((dependents, fixture) => {
        if (fixture.dependencies.has(this.reference)) {
          dependents.push(fixture.reference);
        }

        return dependents;
      }, []),
    );
  }

  @Memoize((context: TRequestContext) => context)
  public async load(context: TRequestContext): Promise<NodeValue> {
    const data = this.node.creationInputType.parseValue(
      Object.fromEntries(
        await Promise.all(
          Object.entries(this.data).map(
            async ([componentName, componentValue]) => [
              componentName,
              this.node.edgesByName.has(componentName)
                ? {
                    [EdgeCreationInputAction.CONNECT]:
                      await this.seeding.fixturesByReference
                        .get(componentValue)!
                        .getIdentifier(context),
                  }
                : componentValue,
            ],
          ),
        ),
      ),
      utils.addPath(undefined, this.reference),
    ) as any;

    return this.node
      .getMutationByKey('create-one')
      .execute({ data, selection: this.node.selection }, context) as any;
  }

  @Memoize((context: TRequestContext) => context)
  public async getIdentifier(
    context: TRequestContext,
  ): Promise<UniqueConstraintValue> {
    const nodeValue = await this.load(context);

    return this.node.identifier.parseValue(nodeValue);
  }
}
