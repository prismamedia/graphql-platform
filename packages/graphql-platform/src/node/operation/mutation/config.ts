import type * as utils from '@prismamedia/graphql-platform-utils';
import type { ConnectorInterface } from '../../../connector-interface.js';
import type { CreationConfig } from './abstract-creation.js';
import type { DeletionConfig } from './abstract-deletion.js';
import type { UpdateConfig } from './abstract-update.js';

export type {
  CreationConfig,
  PostCreateArgs,
  PreCreateArgs,
} from './abstract-creation.js';
export type {
  DeletionConfig,
  PostDeleteArgs,
  PreDeleteArgs,
} from './abstract-deletion.js';
export type {
  PostUpdateArgs,
  PreUpdateArgs,
  UpdateConfig,
} from './abstract-update.js';

export interface MutationConfig<
  TRequestContext extends object,
  TConnector extends ConnectorInterface,
  TContainer extends object,
> {
  [utils.MutationType.CREATION]: CreationConfig<
    TRequestContext,
    TConnector,
    TContainer
  >;
  [utils.MutationType.UPDATE]: UpdateConfig<
    TRequestContext,
    TConnector,
    TContainer
  >;
  [utils.MutationType.DELETION]: DeletionConfig<
    TRequestContext,
    TConnector,
    TContainer
  >;
}
