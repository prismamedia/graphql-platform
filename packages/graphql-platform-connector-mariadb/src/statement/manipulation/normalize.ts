import * as scalars from '@prismamedia/graphql-platform-scalars';
import type * as mariadb from 'mariadb';
import assert from 'node:assert/strict';
import { EOL } from 'node:os';
import { escapeIdentifier, escapeStringValue } from '../../escaping.js';
import type { LeafColumn, Table } from '../../schema.js';
import { StatementKind } from '../kind.js';

export type LeafColumnNormalizer = (expr: string) => string;

export const trimWhitespaces = (expr: string) =>
  `REGEXP_REPLACE(${expr}, '^\\\\s+|\\\\s+$', '')`;

export const normalizeDoubleSpaces = (expr: string) =>
  `REGEXP_REPLACE(${expr}, '\\\\s+', ' ')`;

const normalizeJSON = (
  expr: string,
  type: 'ARRAY' | 'OBJECT',
  onInvalid: string,
) =>
  `IF(${[
    `JSON_VALID(${expr})`,
    `JSON_TYPE(${expr}) = ${escapeStringValue(type)}`,
    `JSON_LENGTH(${expr}) > 0`,
  ].join(' AND ')}, JSON_COMPACT(${expr}), ${onInvalid})`;

export const normalizeJSONArray = (expr: string, isNullable: boolean) =>
  normalizeJSON(
    expr,
    'ARRAY',
    isNullable ? 'NULL' : escapeStringValue(JSON.stringify([])),
  );

export const normalizeJSONObject = (expr: string, isNullable: boolean) =>
  normalizeJSON(
    expr,
    'OBJECT',
    isNullable ? 'NULL' : escapeStringValue(JSON.stringify({})),
  );

export const normalizeDraftJS = (expr: string, isNullable: boolean) =>
  `IF(${[
    `JSON_VALID(${expr})`,
    `JSON_EXISTS(${expr}, '$.blocks')`,
    `JSON_TYPE(JSON_EXTRACT(${expr}, '$.blocks')) != 'NULL'`,
    `JSON_LENGTH(${expr}, '$.blocks') > 0`,
  ].join(' AND ')}, JSON_COMPACT(${expr}), ${
    isNullable
      ? 'NULL'
      : escapeStringValue(JSON.stringify({ entityMap: {}, blocks: [] }))
  })`;

export interface NormalizeStatementConfig {
  /**
   * @see https://mariadb.com/kb/en/ignore/
   */
  ignore?: boolean;

  customize?: (column: LeafColumn, expr?: string) => string | undefined;
}

/**
 * @see https://mariadb.com/kb/en/update/
 */
export class NormalizeStatement implements mariadb.QueryOptions {
  public readonly kind = StatementKind.DATA_MANIPULATION;
  public readonly sql: string;

  public static normalizations(
    table: Table,
    config?: NormalizeStatementConfig,
  ): Map<LeafColumn, string> {
    return new Map<LeafColumn, string>(
      Array.from(
        table.columnsByLeaf.values(),
        (column): [LeafColumn, string] | undefined => {
          let normalizers: Array<LeafColumnNormalizer | undefined> = [];

          switch (column.leaf.type) {
            case scalars.GraphQLNonEmptyString:
              normalizers = [
                column.isNullable()
                  ? (expr) => `NULLIF(${expr}, '')`
                  : undefined,
              ];
              break;

            case scalars.GraphQLEmailAddress:
            case scalars.GraphQLNonEmptyTrimmedString:
            case scalars.GraphQLURL:
              normalizers = [
                trimWhitespaces,
                column.isNullable()
                  ? (expr) => `NULLIF(${expr}, '')`
                  : undefined,
              ];
              break;

            case scalars.GraphQLJSONArray:
              normalizers = [
                (expr) => normalizeJSONArray(expr, column.isNullable()),
              ];
              break;

            case scalars.GraphQLJSONObject:
              normalizers = [
                (expr) => normalizeJSONObject(expr, column.isNullable()),
              ];
              break;

            case scalars.GraphQLDraftJS:
              normalizers = [
                (expr) => normalizeDraftJS(expr, column.isNullable()),
              ];
              break;
          }

          const initialValue = escapeIdentifier(column.name);

          let expr: string | undefined = normalizers.reduce(
            (expr, normalizer) => normalizer?.(expr) ?? expr,
            initialValue,
          );

          config?.customize && (expr = config.customize(column, expr));

          return expr && expr !== initialValue ? [column, expr] : undefined;
        },
      ).filter<[LeafColumn, string]>(Boolean as any),
    );
  }

  public constructor(
    public readonly table: Table,
    config?: NormalizeStatementConfig,
  ) {
    const normalizationsByColumn = (
      this.constructor as typeof NormalizeStatement
    ).normalizations(table, config);

    assert(
      normalizationsByColumn.size > 0,
      `No normalization to perform on "${table.name}"`,
    );

    this.sql = [
      [
        'UPDATE',
        config?.ignore === true && 'IGNORE',
        escapeIdentifier(table.name),
      ]
        .filter(Boolean)
        .join(' '),
      'SET',
      Array.from(
        normalizationsByColumn,
        ([column, expr]) => `  ${escapeIdentifier(column.name)} = ${expr}`,
      ).join(`,${EOL}`),
    ]
      .filter(Boolean)
      .join(EOL);
  }
}
