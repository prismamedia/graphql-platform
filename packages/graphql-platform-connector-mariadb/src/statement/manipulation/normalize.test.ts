import { afterAll, beforeAll, describe, expect, it } from '@jest/globals';
import {
  myAdminContext,
  nodeNames,
} from '@prismamedia/graphql-platform/__tests__/config.js';
import * as fixtures from '@prismamedia/graphql-platform/__tests__/fixture.js';
import { createMyGP, type MyGP } from '../../__tests__/config.js';
import { NormalizeStatement } from './normalize.js';

describe('Normalize statement', () => {
  let gp: MyGP;

  beforeAll(async () => {
    gp = createMyGP(`connector_mariadb_normalize_statement`);

    await gp.connector.setup();
    await gp.seed(myAdminContext, fixtures.constant);
  });

  afterAll(() => gp.connector.teardown());

  it.each(nodeNames)(
    'generates "normalize" statement for "%s"',
    async (nodeName) => {
      const table = gp.connector.schema.getTableByNode(nodeName);

      if (NormalizeStatement.normalizations(table).size) {
        const statement = new NormalizeStatement(table);

        expect(statement.sql).toMatchSnapshot();

        await expect(
          gp.connector.executeStatement(statement),
        ).resolves.toMatchSnapshot();
      } else {
        expect(
          () => new NormalizeStatement(table),
        ).toThrowErrorMatchingSnapshot();
      }
    },
  );
});
