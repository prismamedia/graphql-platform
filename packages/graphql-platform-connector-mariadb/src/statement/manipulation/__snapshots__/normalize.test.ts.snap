// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Normalize statement generates "normalize" statement for "Article" 1`] = `
"UPDATE \`articles\`
SET
  \`title\` = REGEXP_REPLACE(REGEXP_REPLACE(\`title\`, '^\\\\s+|\\\\s+$', ''), '\\\\s+', ' '),
  \`slug\` = REGEXP_REPLACE(REGEXP_REPLACE(\`slug\`, '^\\\\s+|\\\\s+$', ''), '\\\\s+', ' '),
  \`body\` = IF(JSON_VALID(\`body\`) AND JSON_EXISTS(\`body\`, '$.blocks') AND JSON_TYPE(JSON_EXTRACT(\`body\`, '$.blocks')) != 'NULL' AND JSON_LENGTH(\`body\`, '$.blocks') > 0, JSON_COMPACT(\`body\`), NULL),
  \`metas\` = IF(JSON_VALID(\`metas\`) AND JSON_TYPE(\`metas\`) = 'OBJECT' AND JSON_LENGTH(\`metas\`) > 0, JSON_COMPACT(\`metas\`), NULL),
  \`machine_tags\` = IF(JSON_VALID(\`machine_tags\`) AND JSON_TYPE(\`machine_tags\`) = 'ARRAY' AND JSON_LENGTH(\`machine_tags\`) > 0, JSON_COMPACT(\`machine_tags\`), NULL)"
`;

exports[`Normalize statement generates "normalize" statement for "Article" 2`] = `
OkPacket {
  "affectedRows": 6,
  "insertId": 0n,
  "warningStatus": 0,
}
`;

exports[`Normalize statement generates "normalize" statement for "ArticleExtension" 1`] = `
"UPDATE \`article_extensions\`
SET
  \`source\` = NULLIF(REGEXP_REPLACE(REGEXP_REPLACE(\`source\`, '^\\\\s+|\\\\s+$', ''), '\\\\s+', ' '), '')"
`;

exports[`Normalize statement generates "normalize" statement for "ArticleExtension" 2`] = `
OkPacket {
  "affectedRows": 1,
  "insertId": 0n,
  "warningStatus": 0,
}
`;

exports[`Normalize statement generates "normalize" statement for "ArticleTag" 1`] = `"No normalization to perform on "article_tags""`;

exports[`Normalize statement generates "normalize" statement for "ArticleTagModeration" 1`] = `
"UPDATE \`article_tag_moderations\`
SET
  \`moderation\` = NULLIF(REGEXP_REPLACE(REGEXP_REPLACE(\`moderation\`, '^\\\\s+|\\\\s+$', ''), '\\\\s+', ' '), '')"
`;

exports[`Normalize statement generates "normalize" statement for "ArticleTagModeration" 2`] = `
OkPacket {
  "affectedRows": 3,
  "insertId": 0n,
  "warningStatus": 0,
}
`;

exports[`Normalize statement generates "normalize" statement for "Category" 1`] = `
"UPDATE \`categories\`
SET
  \`title\` = REGEXP_REPLACE(REGEXP_REPLACE(\`title\`, '^\\\\s+|\\\\s+$', ''), '\\\\s+', ' '),
  \`slug\` = REGEXP_REPLACE(REGEXP_REPLACE(\`slug\`, '^\\\\s+|\\\\s+$', ''), '\\\\s+', ' ')"
`;

exports[`Normalize statement generates "normalize" statement for "Category" 2`] = `
OkPacket {
  "affectedRows": 3,
  "insertId": 0n,
  "warningStatus": 0,
}
`;

exports[`Normalize statement generates "normalize" statement for "Log" 1`] = `
"UPDATE \`logs\`
SET
  \`message\` = REGEXP_REPLACE(REGEXP_REPLACE(\`message\`, '^\\\\s+|\\\\s+$', ''), '\\\\s+', ' ')"
`;

exports[`Normalize statement generates "normalize" statement for "Log" 2`] = `
OkPacket {
  "affectedRows": 0,
  "insertId": 0n,
  "warningStatus": 0,
}
`;

exports[`Normalize statement generates "normalize" statement for "Tag" 1`] = `
"UPDATE \`tags\`
SET
  \`title\` = REGEXP_REPLACE(REGEXP_REPLACE(\`title\`, '^\\\\s+|\\\\s+$', ''), '\\\\s+', ' '),
  \`slug\` = REGEXP_REPLACE(REGEXP_REPLACE(\`slug\`, '^\\\\s+|\\\\s+$', ''), '\\\\s+', ' ')"
`;

exports[`Normalize statement generates "normalize" statement for "Tag" 2`] = `
OkPacket {
  "affectedRows": 3,
  "insertId": 0n,
  "warningStatus": 0,
}
`;

exports[`Normalize statement generates "normalize" statement for "User" 1`] = `
"UPDATE \`users\`
SET
  \`username\` = REGEXP_REPLACE(REGEXP_REPLACE(\`username\`, '^\\\\s+|\\\\s+$', ''), '\\\\s+', ' ')"
`;

exports[`Normalize statement generates "normalize" statement for "User" 2`] = `
OkPacket {
  "affectedRows": 2,
  "insertId": 0n,
  "warningStatus": 0,
}
`;

exports[`Normalize statement generates "normalize" statement for "UserProfile" 1`] = `"No normalization to perform on "user_profiles""`;
